front/pages/api/w/[wId]/vaults/[vId]/data_sources/index.ts

```diff
@@ -0,0 +1,637 @@
+        data_source_error: dustDataSource.error,
+      },
+    });
+  }
+
+  const dataSource = await DataSourceResource.makeNew(
+    {
+      assistantDefaultSelected:
+        isConnectorProviderAssistantDefaultSelected(provider),
+      connectorProvider: provider,
+      description: dataSourceDescription,
+      dustAPIProjectId: dustProject.value.project.project_id.toString(),
+      editedByUserId: user.id,
+      name: dataSourceName,
+      workspaceId: owner.id,
+    },
+    vault
+  );
+
+  // For each data source, we create two views:
+  // - One default view in its associated vault
+  // - If the data source resides in the system vault, we also create a custom view in the global vault until vault are released.
+  if (dataSource.vault.isSystem()) {
+    const globalVault = await VaultResource.fetchWorkspaceGlobalVault(auth);
+
+    await DataSourceViewResource.createViewInVaultFromDataSourceIncludingAllDocuments(
+      globalVault,
+      dataSource,
+      "custom"
+    );
+  }
+
+  const dataSourceView =
+    await DataSourceViewResource.createViewInVaultFromDataSourceIncludingAllDocuments(
+      dataSource.vault,
+      dataSource
+    );
+
+  const connectorsAPI = new ConnectorsAPI(
+    config.getConnectorsAPIConfig(),
+    logger
+  );
+
+  const connectorsRes = await connectorsAPI.createConnector(
+    provider,
+    owner.sId,
+    systemAPIKeyRes.value.secret,
+    dataSourceName,
+    connectionId ?? "none",
+    configuration
+  );
+
+  if (connectorsRes.isErr()) {
+    logger.error(
+      {
+        error: connectorsRes.error,
+      },
+      "Failed to create the connector"
+    );
+    await dataSource.delete(auth);
+    const deleteRes = await coreAPI.deleteDataSource({
+      projectId: dustProject.value.project.project_id.toString(),
+      dataSourceName: dustDataSource.value.data_source.data_source_id,
+    });
+    if (deleteRes.isErr()) {
+      logger.error(
+        {
+          error: deleteRes.error,
+        },
+        "Failed to delete the data source"
+      );
+    }
+    return apiError(req, res, {
+      status_code: 500,
+      api_error: {
+        type: "internal_server_error",
+        message: "Failed to create the connector.",
+        connectors_error: connectorsRes.error,
+      },
+    });
+  }
+  const email = auth.user()?.email;
+  if (email && !isDisposableEmailDomain(email)) {
+    void sendUserOperationMessage({
+      logger,
+      message: `${email} \`${dataSource.name}\`  for workspace \`${
+        owner.name
+      }\` sId: \`${owner.sId}\` connectorId: \`${
+        connectorsRes.value.id
+      }\` provider: \`${provider}\` trialing: \`${
+        auth.subscription()?.trialing ? "true" : "false"
+      }\``,
+    });
+  }
+
+  await dataSource.update({
+    connectorId: connectorsRes.value.id,
+  });
+  const dataSourceType = await getDataSource(auth, dataSource.name);
+  if (dataSourceType) {
+    void ServerSideTracking.trackDataSourceCreated({
```

Comment:
Should happen after we send the response?

Reply:
Right but I'd be in favor of moving all non blocking calls after the reply for clarity

Reply:
Especially as they can still throw so they should not be on the critical path of answering to the user

Reply:
Ok I added them after and wrapped them in a try/catch to make sure they can't crash.